src/library/scala/collection/immutable/Map.scala

```diff
@@ -206,11 +206,23 @@ object Map extends MapFactory[Map] {
 
   def empty[K, V]: Map[K, V] = EmptyMap.asInstanceOf[Map[K, V]]
 
-  def from[K, V](it: collection.IterableOnce[(K, V)]): Map[K, V] =
+  def from[K, V](it: IterableOnce[(K, V)]): Map[K, V] =
     it match {
       case it: Iterable[_] if it.isEmpty => empty[K, V]
-      case m: Map[K, V] => m
-      case _ => (newBuilder[K, V] ++= it).result()
+      // Since IterableOnce[(K, V)] launders the variance of K,
+      // identify only our implementations which can be soundly substituted.
+      // For example, the ordering used by sorted maps would fail on widened key type. (scala/bug#12745)
+      // The following type test is not sufficient: case m: Map[K, V] => m
+      case m: HashMap[K, V]     => m
+      case m: ListMap[K, V]     => m
+      case m: Map1[K, V]        => m
+      case m: Map2[K, V]        => m
+      case m: Map3[K, V]        => m
+      case m: Map4[K, V]        => m
+      case m: WithDefault[K, V] => m
+      case m: SeqMap[K, V]      => m
```

Comment:
`WithDefault` and `SeqMap` could have external subclasses.

We could add `SeqMap1-4`, and also `VectorMap`?